File: java/src/main/java/benchmarks/BrainfuckRecursion.java
package benchmarks;

import java.util.*;

public class BrainfuckRecursion extends Benchmark {

    interface Op {}

    static class Inc implements Op {}
    static class Dec implements Op {}
    static class Next implements Op {}
    static class Prev implements Op {}
    static class Print implements Op {}
    static class Loop implements Op {
        Op[] body;
        Loop(Op[] body) {
            this.body = body;
        }
    }

    static class Tape {
        private byte[] tape = new byte[30000];
        private int pos = 0;

        byte get() {
            return tape[pos];
        }
        void inc() {
            tape[pos]++;
        }
        void dec() {
            tape[pos]--;
        }

        void next() {
            pos++;
            if (pos >= tape.length) {
                byte[] newTape = new byte[tape.length + 1];
                System.arraycopy(tape, 0, newTape, 0, tape.length);
                tape = newTape;
            }
        }

        void prev() {
            if (pos > 0) pos--;
        }
    }

    static class Program {
        private final Op[] ops;
        private long result = 0;

        Program(String code) {
            int[] pos = {0};
            List<Op> opsList = parse(code, pos);
            this.ops = opsList.toArray(new Op[0]);
        }

        private List<Op> parse(String code, int[] pos) {
            List<Op> list = new ArrayList<>();

            while (pos[0] < code.length()) {
                char c = code.charAt(pos[0]++);
                switch (c) {
                case '+':
                    list.add(new Inc());
                    break;
                case '-':
                    list.add(new Dec());
                    break;
                case '>':
                    list.add(new Next());
                    break;
                case '<':
                    list.add(new Prev());
                    break;
                case '.':
                    list.add(new Print());
                    break;
                case '[':
                    List<Op> loopList = parse(code, pos);
                    list.add(new Loop(loopList.toArray(new Op[0])));
                    break;
                case ']':
                    return list;
                }
            }
            return list;
        }

        long run() {
            Tape tape = new Tape();
            result = 0;
            execute(ops, tape);
            return result;
        }

        private void execute(Op[] program, Tape tape) {
            for (Op op : program) {
                if (op instanceof Inc) {
                    tape.inc();
                } else if (op instanceof Dec) {
                    tape.dec();
                } else if (op instanceof Next) {
                    tape.next();
                } else if (op instanceof Prev) {
                    tape.prev();
                } else if (op instanceof Print) {
                    result = (result << 2) + (tape.get() & 0xFF);
                } else if (op instanceof Loop) {
                    Loop loop = (Loop) op;
                    while (tape.get() != 0) {
                        execute(loop.body, tape);
                    }
                }
            }
        }
    }

    private String text;
    private long resultVal;
    private String warmupProgram;

    public BrainfuckRecursion() {
        text = Helper.configS(name(), "program");
        warmupProgram = Helper.configS(name(), "warmup_program");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "BrainfuckRecursion";
    }

    private long runProgram(String programText) {
        return new Program(programText).run();
    }

    @Override
    public void warmup() {
        long prepareIters = warmupIterations();
        for (long i = 0; i < prepareIters; i++) {
            runProgram(warmupProgram);
        }
    }

    @Override
    public void run(int iterationId) {
        resultVal += runProgram(text);
    }

    @Override
    public long checksum() {
        return resultVal;
    }
}

File: scala/src/main/scala/benchmarks/BrainfuckRecursion.scala
package benchmarks

class BrainfuckRecursion extends Benchmark:
  private val text: String = Helper.configS(name(), "program")
  private val warmupProgram: String = Helper.configS(name(), "warmup_program")
  private var resultVal: Long = 0L

  override def name(): String = "Brainfuck::Recursion"

  sealed trait Op
  case object Inc extends Op
  case object Dec extends Op
  case object Next extends Op
  case object Prev extends Op
  case object Print extends Op
  case class Loop(body: Array[Op]) extends Op

  class Tape:
    private var tape: Array[Byte] = new Array[Byte](30000)
    private var pos: Int = 0

    def get(): Byte = tape(pos)
    def inc(): Unit = tape(pos) = (tape(pos) + 1).toByte
    def dec(): Unit = tape(pos) = (tape(pos) - 1).toByte
    def prev(): Unit = if pos > 0 then pos -= 1
    def next(): Unit =
      pos += 1
      if pos >= tape.length then
        val newTape = new Array[Byte](tape.length + 1)
        System.arraycopy(tape, 0, newTape, 0, tape.length)
        tape = newTape

  class Program(code: String):
    private val ops: Array[Op] =
      val pos = Array(0)
      val opsList = new java.util.ArrayList[Op]()
      parse(opsList, code, pos).toArray(Array.empty[Op])

    private def parse(ops: java.util.List[Op], code: String, pos: Array[Int]): java.util.List[Op] =
      while pos(0) < code.length do
        val c = code.charAt(pos(0))
        pos(0) += 1

        c match
          case '-' => ops.add(Dec)
          case '+' => ops.add(Inc)
          case '<' => ops.add(Prev)
          case '>' => ops.add(Next)
          case '.' => ops.add(Print)
          case '[' =>
            val loopOps = new java.util.ArrayList[Op]()
            parse(loopOps, code, pos)
            ops.add(Loop(loopOps.toArray(Array.empty[Op])))
          case ']' => return ops
          case _   =>
      ops

    def run(): Long =
      val tape = Tape()
      var result = 0L

      def exec(op: Op): Unit = op match
        case Inc        => tape.inc()
        case Dec        => tape.dec()
        case Next       => tape.next()
        case Prev       => tape.prev()
        case Print      => result = (result << 2) + (tape.get() & 0xff)
        case Loop(body) =>
          while tape.get() != 0 do
            var i = 0
            while i < body.length do
              exec(body(i))
              i += 1

      var i = 0
      while i < ops.length do
        exec(ops(i))
        i += 1

      result & 0xffffffffL

  private def runProgram(programText: String): Long =
    Program(programText).run()

  override def warmup(): Unit =
    val prepareIters = warmupIterations()
    var i = 0L
    while i < prepareIters do
      runProgram(warmupProgram)
      i += 1

  override def run(iterationId: Int): Unit =
    resultVal += runProgram(text)

  override def checksum(): Long = resultVal & 0xffffffffL
